### PR TITLE
fix: close popover when dragging over

### PR DESCRIPTION
**Pull request by ishpaul777**
### Details

This PR aims to close Popover when a file is dragover to resolve visual glitch in  https://github.com/Expensify/App/issues/30153

### Fixed Issues

$ https://github.com/Expensify/App/issues/30153
PROPOSAL: https://github.com/Expensify/App/issues/30153#issuecomment-1774769496

### Tests

- [x] Verify that no errors appear in the JS console

1. Open any chat, open mention suggestion Popover (by typing `@` in composer), Verify when a file is dragover the  the chat screen Popover hides.
2. Open any chat, open emoji suggestion Popover (by typing` :[anyemojiName]` in composer), Verify when a file is dragover the chat screen Popover hides.
3. open context menu popover from any message or LHN chat context menu, , Verify when a file is dragover, the Popover hides.

**Note: When directly dragover a file on suggestions popovers the popovers don't close also the https://github.com/Expensify/App/pull/31154#discussion_r1390948352**

### QA Steps

- [x] Verify that no errors appear in the JS console

1. Open any chat, open mention suggestion Popover (by typing `@` in composer), Verify when a file is dragover the Popover hides.
2. Open any chat, open emoji suggestion Popover (by typing` :[anyemojiName]` in composer), Verify when a file is dragover the Popover hides.
3. Open context menu popover from any message or LHN chat, , Verify when a file is dragover the Popover hides.

**Note: When directly dragover a file on suggestions popovers the popovers don't close https://github.com/Expensify/App/pull/31154#discussion_r1390948352**

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/104348397/9dfe0d4a-cea6-4a61-beff-315e234e8193

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/104348397/3e8b7ee7-098d-46c6-977c-df5e770c5e1d

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/104348397/66cf51e3-9d8c-4bf7-aafb-c4fa01596b47

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/104348397/667a2a9a-7ee1-417d-b542-f4b43c5e7be9

</details>
<details>

<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/104348397/5bbb7e48-4d00-41f8-8f2c-001bcbcf2ac1

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/104348397/3bcb3d50-8e75-4878-8f42-837001c5e80f

</details>

### CODE REVIEW

**Comment by cubuspl42** [comment]
@ishpaul777 

Some formatting issue:

image

File: src/pages/home/report/ReportActionCompose/Suggestions.js
Comment (cubuspl42) [inline src/pages/home/report/ReportActionCompose/Suggestions.js:141]: Could you explain how this block correlates to the one above it? Why do we need both?

File: src/pages/settings/AboutPage/AboutPage.js
Comment (cubuspl42) [reply src/pages/settings/AboutPage/AboutPage.js:67]: While I appreciate the efforts to improve the app, the PR should not contain unrelated changes.

The harsh truth is that we won't be rewarded for fixing unrelated bugs while we take the consequences for introducing any regressions. This doesn't have to be intuitive; it's enough that this process works for Expensify.

Comment (cubuspl42) [reply src/pages/settings/AboutPage/AboutPage.js:67]: I'm not saying that you expected the payment, I'm just reminding you that we will, according to the process, face the consequences of any regressions any unrelated change might cause. It's harsh, but this is how it is.

Yes, please revert it.

File: src/hooks/useDragAndDrop.ts
Comment (cubuspl42) [inline src/hooks/useDragAndDrop.ts:86]: Could we call the `PopoverContext` `close` method here?

File: src/pages/home/report/ReportActionCompose/Suggestions.js
Comment (cubuspl42) [reply src/pages/home/report/ReportActionCompose/Suggestions.js:141]: So, in other words...

- dragging a file _directly_ over the suggestions component **didn't** work before
- and after this PR, it starts working

Again, this is an unrelated change, as I see it.

File: src/hooks/useDragAndDrop.ts
Comment (cubuspl42) [reply src/hooks/useDragAndDrop.ts:86]: `PopoverContext` is `App`-level and it's not likely to change. Still, a conditional call as a defense-in-depth might be considered.

While it adds some coupling, it's an acceptable level in my opinion.

This would remove some duplication and could simplify the overall solution if I don't miss some unexpected consequences of this change.

File: src/pages/home/report/ReportActionCompose/Suggestions.js
Comment (cubuspl42) [reply src/pages/home/report/ReportActionCompose/Suggestions.js:141]: > I think with this Pr we are aiming to close popver/suggestion if we dragover a file in app

First and formost, this PR has to _fix the issue_. I approved the idea to fix the issue by closing popovers on drag-enter, but it was still done in the context of _the issue_.

The issue is this visual glitch:

image

The app entered the "drag state" (showing "Drag to upload" text, adding a "glass", being ready for drop, etc.), while we have a **dangling UI element** on top of it.

If the app doesn't enter the "drag state" on `main` right now, I can't see why we were obligated to make it enter the "drag state" after this PR.

Does this make sense?

Comment (cubuspl42) [reply src/pages/home/report/ReportActionCompose/Suggestions.js:141]: While the added code is, arguably, some improvement, it:

- is not strictly required, so it's a regression risk (sorry to be boring, but regressions are **a nightmare**)
- adds code duplication on more than one level, so I couldn't accept it as-is anyway

We need a path forward, and, as I see it, the easiest way is just dropping this block.

File: src/components/PopoverProvider/index.tsx
Comment (cubuspl42) [inline src/components/PopoverProvider/index.tsx:105]: Do we need this block in the context of the new call?

File: src/hooks/useDragAndDrop.ts
Comment (cubuspl42) [inline src/hooks/useDragAndDrop.ts:89]: (I'm talking about this call)

File: src/pages/home/ReportScreen.js
Comment (cubuspl42) [inline src/pages/home/ReportScreen.js:289]: Do we need this block in the context [of the new call](https://github.com/Expensify/App/pull/31154/files#r1392491208)?

File: src/components/PopoverProvider/index.tsx
Comment (cubuspl42) [reply src/components/PopoverProvider/index.tsx:105]: Again, does dragging a file over LHN trigger the "ready-to-drop" behavior on `main`? I don't think it does. And it cannot cause the visual glitch.

If the drag gesture is "discarded" by the app (when we decide not to handle it), closing any popovers is unnecessary. From our perspective, it's kind of like the drag gesture is not present at all.

We aim to make the issue disappear with minimal, reasonable changes (unless a non-minimal solution is cleaner / better / more maintainable).

File: src/components/DragAndDrop/NoDropZone/index.js
Comment (cubuspl42) [inline src/components/DragAndDrop/NoDropZone/index.js:18]: If we don't change the file content, we shouldn't affect its whitespace.

File: src/hooks/useDragAndDrop.ts
Comment (cubuspl42) [inline src/hooks/useDragAndDrop.ts:30]: Do you think it would be easy to add `onDragEnter`, symmetrical to `onDrop`?

I'm wondering if this would allow us to reduce the boilerplate in...

```js
        if (prevIsDraggingOver === isDraggingOver) {
            return;
        }
        if (isDraggingOver) {
        
        // ...
```

File: src/hooks/useDragAndDrop.ts
Comment (cubuspl42) [inline src/hooks/useDragAndDrop.ts:34]: Do we ever set this to `false`?

File: src/hooks/useDragAndDrop.ts
Comment (cubuspl42) [reply src/hooks/useDragAndDrop.ts:30]: Oh, `Suggestions` uses `useContext(DragAndDropContext);`... I'm still not convinced by it. I'll think about this.

**Comment by cubuspl42** [comment]
Also, we have some conflicts that need fixing

File: src/hooks/useDragAndDrop.ts
Comment (cubuspl42) [reply src/hooks/useDragAndDrop.ts:30]: For now let's consider this thread "on hold"

File: src/hooks/useDragAndDrop.ts
Comment (cubuspl42) [reply src/hooks/useDragAndDrop.ts:35]: Then I'd say to remove this parameter in the YAGNI style. It can be easily added when needed.

File: src/components/PopoverProvider/index.tsx
Comment (cubuspl42) [reply src/components/PopoverProvider/index.tsx:105]: > it may be seen as inconsistency and considered to be a regression from this PR.

Hm, I understand. I'll re-test to ensure we're on the same page here.

**Comment by cubuspl42** [comment]
@ishpaul777 Would you merge `main` to see if that resolves the "TypeScript Checks"?

**Comment by cubuspl42** [comment]
This doesn't fail in other PRs 🤔

File: src/components/PopoverProvider/index.tsx
Comment (cubuspl42) [reply src/components/PopoverProvider/index.tsx:105]: > unless we are not aiming to closepopover when file is dragover on LHN or Popover itself

We are not. I gave this a thought, and if this is the only reason to include this block, I insist of removing this. After this gets a C+ approval (i.e. mine), you'll have a chance to add a note about this in the PR, and Joel will take this into consideration when giving his final review. 

> but considering **we do close the popover on RHP** we should also close on LHN for sake of consistency

What do you mean by this? Would you post a video of when we "close the popover on RHP" on `main`?

Comment (cubuspl42) [reply src/components/PopoverProvider/index.tsx:105]: Please...
- either provide a video of the undesired (from your perspective) behavior which happens when we delete this block
  - ...I will then decide whether I agree that the behavior on the video is indeed undesired
- ...or remove this block

Comment (cubuspl42) [reply src/components/PopoverProvider/index.tsx]: I tried to remove this myself and reproduce the issue with the RHP and wasn't able to.

Comment (cubuspl42) [reply src/components/PopoverProvider/index.tsx]: If you agree that it solves our original problem well and addresses the RHP-related concern, feel free to cherry-pick this commit.

File: src/pages/home/report/ReportActionCompose/Suggestions.js
Comment (cubuspl42) [inline src/pages/home/report/ReportActionCompose/Suggestions.js:124]: Could this be simplified to...

```ts
if (!prevIsDraggingOver && isDraggingOver) {
    updateShouldShowSuggestionMenuToFalse();
}
```

File: src/hooks/useDragAndDrop.ts
Comment (cubuspl42) [reply src/hooks/useDragAndDrop.ts:31]: Let's do nothing here. We have to merge this PR some day.

**Comment by cubuspl42** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/2590174/60a811ee-6230-4c90-b832-912357a5fcf9

</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/2590174/44352b5a-98eb-4662-b315-65d1293b43b2

</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Review by jakub-trzebiatowski — CHANGES_REQUESTED** [review]
Please merge `main`

**Review by jakub-trzebiatowski — APPROVED** [review]

**Review by joelbettner — APPROVED** [review]

**Pull request merged**
